mqtt/policy/src/core/builder.rs

```diff
@@ -0,0 +1,706 @@
+            // between identities and identity substitutions.
+            static_ids.insert(identity, static_ops);
+            variable_ids.insert(identity, variable_ops);
+        }
+    }
+
+    (static_ids, variable_ids)
+}
+
+fn process_operations(statement: &Statement20201030) -> (Operations, Operations) {
+    let mut static_ops = Operations::new();
+    let mut variable_ops = Operations::new();
+    for operation in &statement.operations {
+        let (static_res, variable_res) = process_resources(&statement);
+
+        if is_variable_rule(operation) {
+            // if current operation has variables,
+            // then the whole resource subtree need
+            // to be cloned into variables tree.
+            let mut all = static_res.clone();
+            all.merge(variable_res);
+            variable_ops.insert(operation, all);
+        } else {
+            // else, divide static resources and variable resources
+            // between static operations and variable operation.
+            static_ops.insert(operation, static_res);
+            variable_ops.insert(operation, variable_res);
+        }
+    }
+
+    (static_ops, variable_ops)
+}
+
+fn process_resources(statement: &Statement20201030) -> (Resources, Resources) {
+    let mut static_res = Resources::new();
+    let mut variable_res = Resources::new();
+    if statement.resources.is_empty() {
+        static_res.insert("", statement.into());
+    }
+
+    for resource in &statement.resources {
+        // split resources into two static or variable rules:
+        let map = if is_variable_rule(resource) {
+            &mut variable_res
+        } else {
+            &mut static_res
+        };
+
+        map.insert(resource, statement.into());
+    }
+
+    (static_res, variable_res)
+}
+
+fn is_variable_rule(value: &str) -> bool {
+    lazy_static! {
+        static ref VAR_PATTERN: Regex =
+            Regex::new(r#"\{\{[^\{\}]+\}\}"#).expect("failed to create a Regex from pattern");
+    }
+    VAR_PATTERN.is_match(value)
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct PolicyDefinition20201030 {
+    statements: Vec<Statement20201030>,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct Statement20201030 {
+    #[serde(default)]
+    order: usize,
+    #[serde(default)]
+    description: String,
+    effect: Effect20201030,
+    identities: Vec<String>,
+    operations: Vec<String>,
+    #[serde(default)]
+    resources: Vec<String>,
+}
+
+#[derive(Deserialize, Copy, Clone)]
+#[serde(rename_all = "camelCase")]
+enum Effect20201030 {
+    Allow,
+    Deny,
+}
+
+impl Into<EffectOrd> for &Statement20201030 {
+    fn into(self) -> EffectOrd {
+        match self.effect {
+            Effect20201030::Allow => EffectOrd::new(CoreEffect::Allow, self.order),
+            Effect20201030::Deny => EffectOrd::new(CoreEffect::Deny, self.order),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Why not put the tests in a separate file?

Reply:
it is rust practice to keep unit tests in the same file, as a submodule.

You can add tests in a separate file - but those will be integration tests for the crate.